<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>A Bulk Experiment • disize</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="A Bulk Experiment">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">disize</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.20</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/disize.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/a-bulk-experiment.html">A Bulk Experiment</a></li>
    <li><a class="dropdown-item" href="../articles/derivation.html">Derivation of disize</a></li>
    <li><a class="dropdown-item" href="../articles/single-cell-experiment.html">Running disize on a Single-cell Experiment</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>A Bulk Experiment</h1>
            
      

      <div class="d-none name"><code>a-bulk-experiment.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="dependencies">Dependencies<a class="anchor" aria-label="anchor" href="#dependencies"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://toddmccready.github.io/disize/">disize</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://jeroen.r-universe.dev/curl" class="external-link">curl</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Using libcurl 8.5.0 with OpenSSL/3.0.13</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://henrikbengtsson.github.io/R.utils/" class="external-link">R.utils</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: R.oo</span></span>
<span><span class="co">#&gt; Loading required package: R.methodsS3</span></span>
<span><span class="co">#&gt; R.methodsS3 v1.8.2 (2022-06-13 22:00:14 UTC) successfully loaded. See ?R.methodsS3 for help.</span></span>
<span><span class="co">#&gt; R.oo v1.27.1 (2025-05-02 21:00:05 UTC) successfully loaded. See ?R.oo for help.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'R.oo'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:R.methodsS3':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     throw</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:methods':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     getClasses, getMethods</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     attach, detach, load, save</span></span>
<span><span class="co">#&gt; R.utils v2.13.0 (2025-02-24 21:20:02 UTC) successfully loaded. See ?R.utils for help.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'R.utils'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:utils':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     timestamp</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     cat, commandArgs, getOption, isOpen, nullfile, parse, use, warnings</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="the-dataset">The dataset<a class="anchor" aria-label="anchor" href="#the-dataset"></a>
</h2>
<p>This dataset consists of a purified macrophage subtype(“Mac2”,
supposedly in an ‘activated’ state) that has been partitioned into four
groups exposed to different conditions. The authors offer this
information on how the samples were processed:</p>
<blockquote>
<p>The sorted Mac2 cells were divided into four groups and stimulated at
the 3-h time point with the same concentrations as previously described.
Then, the RNA was extracted using the RNeasy Plus Micro Kit as per
manufacturer instructions. Poly(A)mRNA was isolated using mRNA Capture
Beads 2.0 (Yeasen Cat.12629ES, CHN) with two rounds of purification,
followed by RNA fragmentation with magnesium ions at 94°C (Yeasen
Cat.12340ES97, CHN). RNA sequencing library preparation was performed
using the TruSeq RNA Library Prep Kit v2 (Illumina). Sequencing was
carried out as paired-end 2×150 bp (PE150) on an Illumina Novaseq™ X
Plus (LC-Bio Technologies).</p>
</blockquote>
<p>The TruSeq RNA Library Prep Kit involves “tagging” transcripts with
barcodes that identify distinct <em>samples</em>, allowing all prepared
cDNA libraries to be pooled together before sequencing. Since
batch-effects are usually attributed to separate sequencing runs, then
we expect very small batch-effects to be present in this dataset if we
define a “batch” as the unit subjected to RNA extraction(and all further
processing).</p>
<p>Let us test this hypothesis:</p>
</div>
<div class="section level2">
<h2 id="downloading-the-data">Downloading the data<a class="anchor" aria-label="anchor" href="#downloading-the-data"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Download counts and construct metadata</span></span>
<span><span class="va">counts_path</span> <span class="op">&lt;-</span> <span class="fu">curl</span><span class="fu">::</span><span class="fu"><a href="https://jeroen.r-universe.dev/curl/reference/curl_download.html" class="external-link">curl_download</a></span><span class="op">(</span></span>
<span>    url <span class="op">=</span> <span class="st">"https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE273924&amp;format=file&amp;file=GSE273924%5Fraw%5Fcounts%2Etsv%2Egz"</span>,</span>
<span>    destfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"/counts.tsv.gz"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="va">counts_path</span><span class="op">)</span></span>
<span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    <span class="st">"sample_id"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>,</span>
<span>    <span class="st">"condition"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span>, <span class="st">"lps"</span>, <span class="st">"nelps"</span>, <span class="st">"ne"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Coerce to formatted matrix</span></span>
<span><span class="va">gene_names</span> <span class="op">&lt;-</span> <span class="va">counts</span><span class="op">$</span><span class="va">gene_id</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">counts</span><span class="op">[</span>,<span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">gene_names</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="running-disize">Running <code>disize</code><a class="anchor" aria-label="anchor" href="#running-disize"></a>
</h2>
<p>The <code>metadata</code> contains the information for the
experimental design:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></span>
<span><span class="co">#&gt;    sample_id condition</span></span>
<span><span class="co">#&gt; 1      Ctrl1   control</span></span>
<span><span class="co">#&gt; 2      Ctrl2   control</span></span>
<span><span class="co">#&gt; 3      Ctrl3   control</span></span>
<span><span class="co">#&gt; 4       LPS1       lps</span></span>
<span><span class="co">#&gt; 5       LPS2       lps</span></span>
<span><span class="co">#&gt; 6       LPS3       lps</span></span>
<span><span class="co">#&gt; 7     NELPS1     nelps</span></span>
<span><span class="co">#&gt; 8     NELPS2     nelps</span></span>
<span><span class="co">#&gt; 9     NELPS3     nelps</span></span>
<span><span class="co">#&gt; 10       NE1        ne</span></span>
<span><span class="co">#&gt; 11       NE2        ne</span></span>
<span><span class="co">#&gt; 12       NE3        ne</span></span></code></pre></div>
<p>For this dataset, the study was primarily interested in the effect of
<code>condition</code> on expression, thus the formula we would input
into <code>disize</code> is:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">design_formula</span> <span class="op">&lt;-</span> <span class="op">~</span> <span class="va">condition</span></span></code></pre></div>
<p>We can finally run <code>disize</code> to get the estimated size
factors:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">size_factors</span> <span class="op">&lt;-</span> <span class="fu">disize</span><span class="fu">::</span><span class="fu"><a href="../reference/disize.html">disize</a></span><span class="op">(</span></span>
<span>    <span class="va">design_formula</span>,</span>
<span>    <span class="va">counts</span>,</span>
<span>    <span class="va">metadata</span>,</span>
<span>    batch_name <span class="op">=</span> <span class="st">"sample_id"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Formatting data...</span></span>
<span><span class="co">#&gt; ar: creating stan/lib/stan_math/lib/sundials_6.1.1/lib/libsundials_nvecserial.a</span></span>
<span><span class="co">#&gt; ar: creating stan/lib/stan_math/lib/sundials_6.1.1/lib/libsundials_cvodes.a</span></span>
<span><span class="co">#&gt; ar: creating stan/lib/stan_math/lib/sundials_6.1.1/lib/libsundials_idas.a</span></span>
<span><span class="co">#&gt; ar: creating stan/lib/stan_math/lib/sundials_6.1.1/lib/libsundials_kinsol.a</span></span>
<span><span class="co">#&gt; /home/runner/.cmdstan/cmdstan-2.36.0/stan/lib/stan_math/lib/tbb_2020.3/build/Makefile.tbb:28: CONFIG: cfg=release arch=intel64 compiler=gcc target=linux runtime=cc13.3.0_libc2.39_kernel6.11.0</span></span>
<span><span class="co">#&gt; In file included from ../tbb_2020.3/src/tbb/concurrent_hash_map.cpp:17:</span></span>
<span><span class="co">#&gt; ../tbb_2020.3/include/tbb/concurrent_hash_map.h:347:23: warning: ‘template&lt;class _Category, class _Tp, class _Distance, class _Pointer, class _Reference&gt; struct std::iterator’ is deprecated [-Wdeprecated-declarations]</span></span>
<span><span class="co">#&gt;   347 |         : public std::iterator&lt;std::forward_iterator_tag,Value&gt;</span></span>
<span><span class="co">#&gt;       |                       ^~~~~~~~</span></span>
<span><span class="co">#&gt; In file included from /usr/include/c++/13/bits/stl_construct.h:61,</span></span>
<span><span class="co">#&gt;                  from /usr/include/c++/13/bits/stl_tempbuf.h:61,</span></span>
<span><span class="co">#&gt;                  from /usr/include/c++/13/memory:66,</span></span>
<span><span class="co">#&gt;                  from ../tbb_2020.3/include/tbb/tbb_stddef.h:452,</span></span>
<span><span class="co">#&gt;                  from ../tbb_2020.3/include/tbb/concurrent_hash_map.h:23:</span></span>
<span><span class="co">#&gt; /usr/include/c++/13/bits/stl_iterator_base_types.h:127:34: note: declared here</span></span>
<span><span class="co">#&gt;   127 |     struct _GLIBCXX17_DEPRECATED iterator</span></span>
<span><span class="co">#&gt;       |                                  ^~~~~~~~</span></span>
<span><span class="co">#&gt; cc1plus: note: unrecognized command-line option ‘-Wno-unknown-warning-option’ may have been intended to silence earlier diagnostics</span></span>
<span><span class="co">#&gt; In file included from ../tbb_2020.3/src/tbb/concurrent_queue.cpp:22:</span></span>
<span><span class="co">#&gt; ../tbb_2020.3/include/tbb/internal/_concurrent_queue_impl.h:749:21: warning: ‘template&lt;class _Category, class _Tp, class _Distance, class _Pointer, class _Reference&gt; struct std::iterator’ is deprecated [-Wdeprecated-declarations]</span></span>
<span><span class="co">#&gt;   749 |         public std::iterator&lt;std::forward_iterator_tag,Value&gt; {</span></span>
<span><span class="co">#&gt;       |                     ^~~~~~~~</span></span>
<span><span class="co">#&gt; In file included from /usr/include/c++/13/bits/stl_construct.h:61,</span></span>
<span><span class="co">#&gt;                  from /usr/include/c++/13/bits/stl_tempbuf.h:61,</span></span>
<span><span class="co">#&gt;                  from /usr/include/c++/13/memory:66,</span></span>
<span><span class="co">#&gt;                  from ../tbb_2020.3/include/tbb/tbb_stddef.h:452,</span></span>
<span><span class="co">#&gt;                  from ../tbb_2020.3/src/tbb/concurrent_queue.cpp:17:</span></span>
<span><span class="co">#&gt; /usr/include/c++/13/bits/stl_iterator_base_types.h:127:34: note: declared here</span></span>
<span><span class="co">#&gt;   127 |     struct _GLIBCXX17_DEPRECATED iterator</span></span>
<span><span class="co">#&gt;       |                                  ^~~~~~~~</span></span>
<span><span class="co">#&gt; ../tbb_2020.3/include/tbb/internal/_concurrent_queue_impl.h:1013:21: warning: ‘template&lt;class _Category, class _Tp, class _Distance, class _Pointer, class _Reference&gt; struct std::iterator’ is deprecated [-Wdeprecated-declarations]</span></span>
<span><span class="co">#&gt;  1013 |         public std::iterator&lt;std::forward_iterator_tag,Value&gt; {</span></span>
<span><span class="co">#&gt;       |                     ^~~~~~~~</span></span>
<span><span class="co">#&gt; /usr/include/c++/13/bits/stl_iterator_base_types.h:127:34: note: declared here</span></span>
<span><span class="co">#&gt;   127 |     struct _GLIBCXX17_DEPRECATED iterator</span></span>
<span><span class="co">#&gt;       |                                  ^~~~~~~~</span></span>
<span><span class="co">#&gt; cc1plus: note: unrecognized command-line option ‘-Wno-unknown-warning-option’ may have been intended to silence earlier diagnostics</span></span>
<span><span class="co">#&gt; Estimating size factors... (Max ETA: ~109.2s)</span></span>
<span><span class="co">#&gt; Finised in 95.2s!</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">size_factors</span><span class="op">)</span></span>
<span><span class="co">#&gt;        Ctrl1        Ctrl2        Ctrl3         LPS1         LPS2         LPS3 </span></span>
<span><span class="co">#&gt;  0.082340728  0.003553523 -0.086471855  0.085881138  0.012432106  0.100065963 </span></span>
<span><span class="co">#&gt;          NE1          NE2          NE3       NELPS1       NELPS2       NELPS3 </span></span>
<span><span class="co">#&gt; -0.069011169  0.036273935  0.106796358 -0.162830180 -0.134203900 -0.021313770</span></span></code></pre></div>
<p>Evidently the batch-effect is indeed small across most samples! The
samples “NELPS1” and “NELPS2” seem to have been processed slightly
worse, but otherwise the estimated size factors are approximately the
same (within ~0.1).</p>
<p>We can confirm these estimates by rerunning <code>disize</code> with
a larger <code>n_feats</code>:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">size_factors_2</span> <span class="op">&lt;-</span> <span class="fu">disize</span><span class="fu">::</span><span class="fu"><a href="../reference/disize.html">disize</a></span><span class="op">(</span></span>
<span>    <span class="va">design_formula</span>,</span>
<span>    <span class="va">counts</span>,</span>
<span>    <span class="va">metadata</span>,</span>
<span>    batch_name <span class="op">=</span> <span class="st">"sample_id"</span>,</span>
<span>    n_feats <span class="op">=</span> <span class="fl">15000</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Formatting data...</span></span>
<span><span class="co">#&gt; Estimating size factors... (Max ETA: ~160.6s)</span></span>
<span><span class="co">#&gt; Finised in 139.7s!</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">size_factors_2</span><span class="op">)</span></span>
<span><span class="co">#&gt;        Ctrl1        Ctrl2        Ctrl3         LPS1         LPS2         LPS3 </span></span>
<span><span class="co">#&gt;  0.095303742  0.005168706 -0.083965484  0.073874335 -0.008215053  0.087963097 </span></span>
<span><span class="co">#&gt;          NE1          NE2          NE3       NELPS1       NELPS2       NELPS3 </span></span>
<span><span class="co">#&gt; -0.044352619  0.055490226  0.123255579 -0.181307332 -0.149171063 -0.025856786</span></span></code></pre></div>
<p>Indeed the estimates remain largely the same.</p>
</div>
<div class="section level2">
<h2 id="downstream-analysis">Downstream analysis<a class="anchor" aria-label="anchor" href="#downstream-analysis"></a>
</h2>
<p>WIP</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Todd McCready.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
