<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>An introduction to disize • disize</title>
<!-- katex math --><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous">
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script><script src="../katex-auto.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="An introduction to disize">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">disize</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.21</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/disize.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/derivation.html">Derivation of disize</a></li>
    <li><a class="dropdown-item" href="../articles/single-cell-experiment.html">Running disize on a Single-cell Experiment</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/toddmccready/disize/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>An introduction to disize</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/toddmccready/disize/blob/main/vignettes/disize.Rmd" class="external-link"><code>vignettes/disize.Rmd</code></a></small>
      <div class="d-none name"><code>disize.Rmd</code></div>
    </div>

    
    
<p><code>disize</code> revolves around its only export:
<code><a href="../reference/disize.html">disize::disize</a></code>! It has the following required
arguments:</p>
<ul>
<li><p><code>design_formula</code>: An R formula that specifies the
experimental design. This is essentially describing the model that you
would’ve liked to fit in the <em>absence</em> of batch-effects; it is
the same R formula you would pass to tools like <code>DESeq2</code> or
<code>edgeR</code> for differential expression analysis (including
predictors like <code>condition</code>, <code>sex</code>, etc measured
in your study). All terms used in this formula should be present in
<code>metadata</code>.</p></li>
<li><p><code>counts</code>: A (observation x features) matrix containing
the transcript counts. This can be dense or sparse; an internal coercion
to a dense matrix will be done after subsetting relevant features and
observations.</p></li>
<li><p><code>metadata</code>: A dataframe containing the
observation-level information(with observations as rows).</p></li>
<li><p><code>batch_name</code>: The name of the column in
<code>metadata</code> containing the batch identifier.</p></li>
</ul>
<blockquote>
<p>If the row indices of <code>counts</code> and <code>metadata</code>
do not correspond to the same observation but are named, you can specify
<code>obs_name</code> which will automatically re-order
<code>counts</code> to match the indices in <code>metadata</code>.</p>
</blockquote>
<div class="section level2">
<h2 id="an-example-with-deseq2">An example with DESeq2<a class="anchor" aria-label="anchor" href="#an-example-with-deseq2"></a>
</h2>
<p>To see <code>disize</code> in action, we will be analyzing the
following dataset from the <code>pasilla</code> package with
<code>DESeq2</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pasilla</span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://toddmccready.github.io/disize/">disize</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/thelovelab/DESeq2" class="external-link">DESeq2</a></span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="reading-in-data">Reading in data<a class="anchor" aria-label="anchor" href="#reading-in-data"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Grab filepaths</span></span>
<span><span class="va">counts_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>    <span class="st">"extdata"</span>,</span>
<span>    <span class="st">"pasilla_gene_counts.tsv"</span>,</span>
<span>    package<span class="op">=</span><span class="st">"pasilla"</span>,</span>
<span>    mustWork<span class="op">=</span><span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">metadata_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>    <span class="st">"extdata"</span>,</span>
<span>    <span class="st">"pasilla_sample_annotation.csv"</span>,</span>
<span>    package<span class="op">=</span><span class="st">"pasilla"</span>,</span>
<span>    mustWork<span class="op">=</span><span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Read in data</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="va">counts_path</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, row.names <span class="op">=</span> <span class="st">"gene_id"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="va">metadata_path</span><span class="op">)</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>obs_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"fb"</span>, <span class="st">""</span>, <span class="va">metadata</span><span class="op">$</span><span class="va">file</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">obs_id</span>, <span class="va">condition</span>, <span class="va">type</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html" class="external-link">across</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">condition</span>, <span class="va">type</span><span class="op">)</span>, <span class="va">as.factor</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>        id_1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/interaction.html" class="external-link">interaction</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">$</span><span class="va">type</span>, <span class="va">metadata</span><span class="op">$</span><span class="va">condition</span>, sep <span class="op">=</span> <span class="st">":"</span><span class="op">)</span>,</span>
<span>        id_2 <span class="op">=</span> <span class="va">obs_id</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="inspecting-the-study-design">Inspecting the study design<a class="anchor" aria-label="anchor" href="#inspecting-the-study-design"></a>
</h3>
<p>The dataset is derived from a study investigating the effect of
various RNA binding proteins (RBPs) on alternative splicing regulation.
The authors partitioned <em>D. melanogaster</em> cells into two
different conditions: observations with <code>condition = treated</code>
were treated with dsRNAs to knockdown expression via RNAi, while
<code>condition = untreated</code> were left alone as a control.
Additionally, two sets of duplicates were processed in parallel for each
condition with different kits used for library preparation (producing
single-read vs paired-end reads). Thus, a suitable definition of “batch”
might be the interaction between <code>type</code> and
<code>condition</code>. However, it is easy to double-check by making
the batch definition as granular as possible (where each sample denotes
a batch).</p>
<p>The <code>metadata</code> object contains the information from the
experimental design:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></span>
<span><span class="co">#&gt;       obs_id condition        type                  id_1       id_2</span></span>
<span><span class="co">#&gt; 1   treated1   treated single-read   single-read:treated   treated1</span></span>
<span><span class="co">#&gt; 2   treated2   treated  paired-end    paired-end:treated   treated2</span></span>
<span><span class="co">#&gt; 3   treated3   treated  paired-end    paired-end:treated   treated3</span></span>
<span><span class="co">#&gt; 4 untreated1 untreated single-read single-read:untreated untreated1</span></span>
<span><span class="co">#&gt; 5 untreated2 untreated single-read single-read:untreated untreated2</span></span>
<span><span class="co">#&gt; 6 untreated3 untreated  paired-end  paired-end:untreated untreated3</span></span>
<span><span class="co">#&gt; 7 untreated4 untreated  paired-end  paired-end:untreated untreated4</span></span></code></pre></div>
<p>Evidently, one sample was thrown away!</p>
<p>In the absence of batch-effects we are interested in the effects of
<code>condition</code>, so our <code>design_formula</code> should
be:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">design_formula</span> <span class="op">&lt;-</span> <span class="op">~</span> <span class="va">condition</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="estimating-size-factors">Estimating size factors<a class="anchor" aria-label="anchor" href="#estimating-size-factors"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># First batch definition</span></span>
<span><span class="va">size_factors_1</span> <span class="op">&lt;-</span> <span class="fu">disize</span><span class="fu">::</span><span class="fu"><a href="../reference/disize.html">disize</a></span><span class="op">(</span></span>
<span>    <span class="va">design_formula</span>,</span>
<span>    <span class="va">counts</span>,</span>
<span>    <span class="va">metadata</span>,</span>
<span>    batch_name <span class="op">=</span> <span class="st">"id_1"</span>,</span>
<span>    obs_name <span class="op">=</span> <span class="st">"obs_id"</span> <span class="co"># needed to order 'counts' correctly</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Formatting data...</span></span>
<span><span class="co">#&gt; ar: creating stan/lib/stan_math/lib/sundials_6.1.1/lib/libsundials_nvecserial.a</span></span>
<span><span class="co">#&gt; ar: creating stan/lib/stan_math/lib/sundials_6.1.1/lib/libsundials_cvodes.a</span></span>
<span><span class="co">#&gt; ar: creating stan/lib/stan_math/lib/sundials_6.1.1/lib/libsundials_idas.a</span></span>
<span><span class="co">#&gt; ar: creating stan/lib/stan_math/lib/sundials_6.1.1/lib/libsundials_kinsol.a</span></span>
<span><span class="co">#&gt; /home/runner/.cmdstan/cmdstan-2.36.0/stan/lib/stan_math/lib/tbb_2020.3/build/Makefile.tbb:28: CONFIG: cfg=release arch=intel64 compiler=gcc target=linux runtime=cc13.3.0_libc2.39_kernel6.11.0</span></span>
<span><span class="co">#&gt; In file included from ../tbb_2020.3/src/tbb/concurrent_hash_map.cpp:17:</span></span>
<span><span class="co">#&gt; ../tbb_2020.3/include/tbb/concurrent_hash_map.h:347:23: warning: ‘template&lt;class _Category, class _Tp, class _Distance, class _Pointer, class _Reference&gt; struct std::iterator’ is deprecated [-Wdeprecated-declarations]</span></span>
<span><span class="co">#&gt;   347 |         : public std::iterator&lt;std::forward_iterator_tag,Value&gt;</span></span>
<span><span class="co">#&gt;       |                       ^~~~~~~~</span></span>
<span><span class="co">#&gt; In file included from /usr/include/c++/13/bits/stl_construct.h:61,</span></span>
<span><span class="co">#&gt;                  from /usr/include/c++/13/bits/stl_tempbuf.h:61,</span></span>
<span><span class="co">#&gt;                  from /usr/include/c++/13/memory:66,</span></span>
<span><span class="co">#&gt;                  from ../tbb_2020.3/include/tbb/tbb_stddef.h:452,</span></span>
<span><span class="co">#&gt;                  from ../tbb_2020.3/include/tbb/concurrent_hash_map.h:23:</span></span>
<span><span class="co">#&gt; /usr/include/c++/13/bits/stl_iterator_base_types.h:127:34: note: declared here</span></span>
<span><span class="co">#&gt;   127 |     struct _GLIBCXX17_DEPRECATED iterator</span></span>
<span><span class="co">#&gt;       |                                  ^~~~~~~~</span></span>
<span><span class="co">#&gt; cc1plus: note: unrecognized command-line option ‘-Wno-unknown-warning-option’ may have been intended to silence earlier diagnostics</span></span>
<span><span class="co">#&gt; In file included from ../tbb_2020.3/src/tbb/concurrent_queue.cpp:22:</span></span>
<span><span class="co">#&gt; ../tbb_2020.3/include/tbb/internal/_concurrent_queue_impl.h:749:21: warning: ‘template&lt;class _Category, class _Tp, class _Distance, class _Pointer, class _Reference&gt; struct std::iterator’ is deprecated [-Wdeprecated-declarations]</span></span>
<span><span class="co">#&gt;   749 |         public std::iterator&lt;std::forward_iterator_tag,Value&gt; {</span></span>
<span><span class="co">#&gt;       |                     ^~~~~~~~</span></span>
<span><span class="co">#&gt; In file included from /usr/include/c++/13/bits/stl_construct.h:61,</span></span>
<span><span class="co">#&gt;                  from /usr/include/c++/13/bits/stl_tempbuf.h:61,</span></span>
<span><span class="co">#&gt;                  from /usr/include/c++/13/memory:66,</span></span>
<span><span class="co">#&gt;                  from ../tbb_2020.3/include/tbb/tbb_stddef.h:452,</span></span>
<span><span class="co">#&gt;                  from ../tbb_2020.3/src/tbb/concurrent_queue.cpp:17:</span></span>
<span><span class="co">#&gt; /usr/include/c++/13/bits/stl_iterator_base_types.h:127:34: note: declared here</span></span>
<span><span class="co">#&gt;   127 |     struct _GLIBCXX17_DEPRECATED iterator</span></span>
<span><span class="co">#&gt;       |                                  ^~~~~~~~</span></span>
<span><span class="co">#&gt; ../tbb_2020.3/include/tbb/internal/_concurrent_queue_impl.h:1013:21: warning: ‘template&lt;class _Category, class _Tp, class _Distance, class _Pointer, class _Reference&gt; struct std::iterator’ is deprecated [-Wdeprecated-declarations]</span></span>
<span><span class="co">#&gt;  1013 |         public std::iterator&lt;std::forward_iterator_tag,Value&gt; {</span></span>
<span><span class="co">#&gt;       |                     ^~~~~~~~</span></span>
<span><span class="co">#&gt; /usr/include/c++/13/bits/stl_iterator_base_types.h:127:34: note: declared here</span></span>
<span><span class="co">#&gt;   127 |     struct _GLIBCXX17_DEPRECATED iterator</span></span>
<span><span class="co">#&gt;       |                                  ^~~~~~~~</span></span>
<span><span class="co">#&gt; cc1plus: note: unrecognized command-line option ‘-Wno-unknown-warning-option’ may have been intended to silence earlier diagnostics</span></span>
<span><span class="co">#&gt; Estimating size factors... (Max ETA: ~63.7s)</span></span>
<span><span class="co">#&gt; Finised in 60.6s!</span></span>
<span></span>
<span><span class="co"># Second batch definition</span></span>
<span><span class="va">size_factors_2</span> <span class="op">&lt;-</span> <span class="fu">disize</span><span class="fu">::</span><span class="fu"><a href="../reference/disize.html">disize</a></span><span class="op">(</span></span>
<span>    <span class="va">design_formula</span>,</span>
<span>    <span class="va">counts</span>,</span>
<span>    <span class="va">metadata</span>,</span>
<span>    batch_name <span class="op">=</span> <span class="st">"id_2"</span>,</span>
<span>    obs_name <span class="op">=</span> <span class="st">"obs_id"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Formatting data...</span></span>
<span><span class="co">#&gt; Estimating size factors... (Max ETA: ~71.2s)</span></span>
<span><span class="co">#&gt; Finised in 60.7s!</span></span></code></pre></div>
<div class="section level4">
<h4 id="comparing-batch-definitions">Comparing batch definitions<a class="anchor" aria-label="anchor" href="#comparing-batch-definitions"></a>
</h4>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">size_factors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    obs_id <span class="op">=</span> <span class="va">metadata</span><span class="op">$</span><span class="va">obs_id</span>,</span>
<span>    def_1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unname.html" class="external-link">unname</a></span><span class="op">(</span><span class="va">size_factors_1</span><span class="op">[</span><span class="va">metadata</span><span class="op">$</span><span class="va">id_1</span><span class="op">]</span><span class="op">)</span>,</span>
<span>    def_2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unname.html" class="external-link">unname</a></span><span class="op">(</span><span class="va">size_factors_2</span><span class="op">[</span><span class="va">metadata</span><span class="op">$</span><span class="va">id_2</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">size_factors</span><span class="op">)</span></span>
<span><span class="co">#&gt;       obs_id      def_1      def_2</span></span>
<span><span class="co">#&gt; 1   treated1  0.3576840  0.4337995</span></span>
<span><span class="co">#&gt; 2   treated2 -0.3341442 -0.3501711</span></span>
<span><span class="co">#&gt; 3   treated3 -0.3341442 -0.2633270</span></span>
<span><span class="co">#&gt; 4 untreated1  0.2187259  0.0424296</span></span>
<span><span class="co">#&gt; 5 untreated2  0.2187259  0.5012903</span></span>
<span><span class="co">#&gt; 6 untreated3 -0.4950414 -0.5118620</span></span>
<span><span class="co">#&gt; 7 untreated4 -0.4950414 -0.3706849</span></span></code></pre></div>
<p>Although there are some similarities, there is enough of a difference
for some observations that using the more granular definition may bring
some needed precision.</p>
</div>
</div>
<div class="section level3">
<h3 id="running-deseq">Running DESeq<a class="anchor" aria-label="anchor" href="#running-deseq"></a>
</h3>
<p>Once the size factors are inserted into the <code>DESeqDataSet</code>
object, the analysis then proceeds normally.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Constructing DESeqDataSet object</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html" class="external-link">DESeqDataSetFromMatrix</a></span><span class="op">(</span></span>
<span>    countData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">counts</span><span class="op">[</span><span class="va">metadata</span><span class="op">$</span><span class="va">obs_id</span>, <span class="op">]</span><span class="op">)</span>,</span>
<span>    colData <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    design <span class="op">=</span> <span class="va">design_formula</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">sizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="va">size_factors_2</span><span class="op">)</span> <span class="co"># Insert size factors</span></span>
<span></span>
<span><span class="co"># Run analysis</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html" class="external-link">DESeq</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span><span class="co">#&gt; using pre-existing size factors</span></span>
<span><span class="co">#&gt; estimating dispersions</span></span>
<span><span class="co">#&gt; gene-wise dispersion estimates</span></span>
<span><span class="co">#&gt; mean-dispersion relationship</span></span>
<span><span class="co">#&gt; final dispersion estimates</span></span>
<span><span class="co">#&gt; fitting model and testing</span></span>
<span></span>
<span><span class="co"># Print results</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html" class="external-link">results</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; log2 fold change (MLE): condition untreated vs treated </span></span>
<span><span class="co">#&gt; Wald test p-value: condition untreated vs treated </span></span>
<span><span class="co">#&gt; DataFrame with 14599 rows and 6 columns</span></span>
<span><span class="co">#&gt;                baseMean log2FoldChange     lfcSE       stat    pvalue      padj</span></span>
<span><span class="co">#&gt;               &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; FBgn0000003    0.185893     -1.0167939  3.804518 -0.2672596 0.7892693        NA</span></span>
<span><span class="co">#&gt; FBgn0000008  102.946483      0.0152384  0.221325  0.0688507 0.9451085  0.990455</span></span>
<span><span class="co">#&gt; FBgn0000014    1.141458      0.5268319  2.114015  0.2492092 0.8031990        NA</span></span>
<span><span class="co">#&gt; FBgn0000015    0.917912      1.8982548  2.060040  0.9214650 0.3568077        NA</span></span>
<span><span class="co">#&gt; FBgn0000017 4711.140900      0.2564267  0.126126  2.0330989 0.0420425  0.234177</span></span>
<span><span class="co">#&gt; ...                 ...            ...       ...        ...       ...       ...</span></span>
<span><span class="co">#&gt; FBgn0261571 9.25775e-02    -0.87944028  3.809549 -0.2308515  0.817430        NA</span></span>
<span><span class="co">#&gt; FBgn0261572 6.71659e+00     0.97726858  0.762876  1.2810322  0.200182  0.579617</span></span>
<span><span class="co">#&gt; FBgn0261573 2.42519e+03     0.00375871  0.115481  0.0325484  0.974035  0.995694</span></span>
<span><span class="co">#&gt; FBgn0261574 5.25085e+03     0.00508216  0.190924  0.0266188  0.978764  0.997642</span></span>
<span><span class="co">#&gt; FBgn0261575 1.15072e+01    -0.13557712  0.928857 -0.1459612  0.883952  0.973689</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Todd McCready.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
