<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>An introduction to disize • disize</title>
<!-- katex math --><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous">
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script><script src="../katex-auto.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="An introduction to disize">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">disize</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.29</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/disize.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/benchmarking.html">Benchmarking against `DESeq2` and `edgeR`</a></li>
    <li><a class="dropdown-item" href="../articles/comparing-defs.html">Comparing batch definitions</a></li>
    <li><a class="dropdown-item" href="../articles/implementation.html">Implementation of disize</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/toddmccready/disize/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>An introduction to disize</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/toddmccready/disize/blob/main/vignettes/disize.Rmd" class="external-link"><code>vignettes/disize.Rmd</code></a></small>
      <div class="d-none name"><code>disize.Rmd</code></div>
    </div>

    
    
<p><code>disize</code> revolves around its only export:
<code><a href="../reference/disize.html">disize::disize</a></code>! It has the following required
arguments:</p>
<ul>
<li><p><code>design_formula</code>: An R formula that specifies the
experimental design. This is essentially describing the model that you
would’ve liked to fit in the <em>absence</em> of batch-effects; it is
the same R formula you would pass to tools like <code>DESeq2</code> or
<code>edgeR</code> for differential expression analysis (including
predictors like <code>condition</code>, <code>sex</code>, etc measured
in your study). All terms used in this formula should be present in
<code>metadata</code>.</p></li>
<li><p><code>counts</code>: A (observation x features) matrix containing
the transcript counts. This can be dense or sparse; an internal coercion
to a dense matrix will be done after subsetting relevant features and
observations.</p></li>
<li><p><code>metadata</code>: A dataframe containing the
observation-level information(with observations as rows).</p></li>
<li><p><code>batch_name</code>: The name of the column in
<code>metadata</code> containing the batch identifier.</p></li>
</ul>
<blockquote>
<p>If the row indices of <code>counts</code> and <code>metadata</code>
do not correspond to the same observation but are named, you can specify
<code>obs_name</code> which will automatically re-order
<code>counts</code> to match the indices in <code>metadata</code>.</p>
</blockquote>
<div class="section level2">
<h2 id="an-example-with-deseq2">An example with DESeq2<a class="anchor" aria-label="anchor" href="#an-example-with-deseq2"></a>
</h2>
<p>To see <code>disize</code> in action, we will be analyzing the
following dataset from Li et al., 2025 with <code>DESeq2</code>.</p>
<p>This dataset consists of a purified macrophage subtype(“Mac2”,
induced in an ‘activated’ state) that has been partitioned into four
groups exposed to different conditions. The authors offer this
information on how the samples were processed:</p>
<blockquote>
<p>The sorted Mac2 cells were divided into four groups and stimulated at
the 3-h time point with the same concentrations as previously described.
Then, the RNA was extracted using the RNeasy Plus Micro Kit as per
manufacturer instructions. Poly(A)mRNA was isolated using mRNA Capture
Beads 2.0 (Yeasen Cat.12629ES, CHN) with two rounds of purification,
followed by RNA fragmentation with magnesium ions at 94°C (Yeasen
Cat.12340ES97, CHN). RNA sequencing library preparation was performed
using the TruSeq RNA Library Prep Kit v2 (Illumina). Sequencing was
carried out as paired-end 2×150 bp (PE150) on an Illumina Novaseq™ X
Plus (LC-Bio Technologies).</p>
</blockquote>
<p>The TruSeq RNA Library Prep Kit involves “tagging” transcripts with
barcodes that identify distinct <em>samples</em>, allowing all prepared
cDNA libraries to be pooled together before sequencing. Since
batch-effects are usually attributed to separate sequencing runs, then
we expect very small batch-effects to be present in this dataset if we
define a “batch” as the unit subjected to RNA extraction (and all
further processing).</p>
<div class="section level3">
<h3 id="dependencies">Dependencies<a class="anchor" aria-label="anchor" href="#dependencies"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://toddmccready.github.io/disize/">disize</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://jeroen.r-universe.dev/curl" class="external-link">curl</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://henrikbengtsson.github.io/R.utils/" class="external-link">R.utils</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="downloading-the-data">Downloading the data<a class="anchor" aria-label="anchor" href="#downloading-the-data"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Download counts and construct metadata</span></span>
<span><span class="va">counts_path</span> <span class="op">&lt;-</span> <span class="fu">curl</span><span class="fu">::</span><span class="fu"><a href="https://jeroen.r-universe.dev/curl/reference/curl_download.html" class="external-link">curl_download</a></span><span class="op">(</span></span>
<span>    url <span class="op">=</span> <span class="st">"https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE273924&amp;format=file&amp;file=GSE273924%5Fraw%5Fcounts%2Etsv%2Egz"</span>,</span>
<span>    destfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"/counts.tsv.gz"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="va">counts_path</span><span class="op">)</span></span>
<span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    <span class="st">"sample_id"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>,</span>
<span>    <span class="st">"condition"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span>, <span class="st">"lps"</span>, <span class="st">"nelps"</span>, <span class="st">"ne"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Coerce to formatted matrix</span></span>
<span><span class="va">gene_names</span> <span class="op">&lt;-</span> <span class="va">counts</span><span class="op">$</span><span class="va">gene_id</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">counts</span><span class="op">[</span>,<span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">gene_names</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="running-disize">Running <code>disize</code><a class="anchor" aria-label="anchor" href="#running-disize"></a>
</h3>
<p>The <code>metadata</code> contains the information for the
experimental design:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></span>
<span><span class="co">#&gt;    sample_id condition</span></span>
<span><span class="co">#&gt; 1      Ctrl1   control</span></span>
<span><span class="co">#&gt; 2      Ctrl2   control</span></span>
<span><span class="co">#&gt; 3      Ctrl3   control</span></span>
<span><span class="co">#&gt; 4       LPS1       lps</span></span>
<span><span class="co">#&gt; 5       LPS2       lps</span></span>
<span><span class="co">#&gt; 6       LPS3       lps</span></span>
<span><span class="co">#&gt; 7     NELPS1     nelps</span></span>
<span><span class="co">#&gt; 8     NELPS2     nelps</span></span>
<span><span class="co">#&gt; 9     NELPS3     nelps</span></span>
<span><span class="co">#&gt; 10       NE1        ne</span></span>
<span><span class="co">#&gt; 11       NE2        ne</span></span>
<span><span class="co">#&gt; 12       NE3        ne</span></span></code></pre></div>
<p>For this dataset, the study was primarily interested in the effect of
<code>condition</code> on expression, thus the formula we would input
into <code>disize</code> is:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">design_formula</span> <span class="op">&lt;-</span> <span class="op">~</span> <span class="va">condition</span></span></code></pre></div>
<p>We can finally run <code>disize</code> to get the estimated size
factors:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">size_factors_1</span> <span class="op">&lt;-</span> <span class="fu">disize</span><span class="fu">::</span><span class="fu"><a href="../reference/disize.html">disize</a></span><span class="op">(</span></span>
<span>    <span class="va">design_formula</span>,</span>
<span>    <span class="va">counts</span>,</span>
<span>    <span class="va">metadata</span>,</span>
<span>    batch_name <span class="op">=</span> <span class="st">"sample_id"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Formatting data...</span></span>
<span><span class="co">#&gt; Estimating size factors... (Max ETA: ~110.7s)</span></span>
<span><span class="co">#&gt; Finised in 96.7s!</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">size_factors_1</span><span class="op">)</span></span>
<span><span class="co">#&gt;        Ctrl1        Ctrl2        Ctrl3         LPS1         LPS2         LPS3 </span></span>
<span><span class="co">#&gt;  0.082340728  0.003553523 -0.086471855  0.085881138  0.012432106  0.100065963 </span></span>
<span><span class="co">#&gt;          NE1          NE2          NE3       NELPS1       NELPS2       NELPS3 </span></span>
<span><span class="co">#&gt; -0.069011169  0.036273935  0.106796358 -0.162830180 -0.134203900 -0.021313770</span></span></code></pre></div>
<p>Evidently the batch-effect is indeed small across most samples! The
samples “NELPS1” and “NELPS2” seem to have been processed slightly
worse, but otherwise the estimated size factors are approximately the
same (within ~0.1).</p>
<p>We can confirm these estimates by rerunning <code>disize</code> with
a larger <code>n_feats</code>:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">size_factors_2</span> <span class="op">&lt;-</span> <span class="fu">disize</span><span class="fu">::</span><span class="fu"><a href="../reference/disize.html">disize</a></span><span class="op">(</span></span>
<span>    <span class="va">design_formula</span>,</span>
<span>    <span class="va">counts</span>,</span>
<span>    <span class="va">metadata</span>,</span>
<span>    batch_name <span class="op">=</span> <span class="st">"sample_id"</span>,</span>
<span>    n_feats <span class="op">=</span> <span class="fl">15000</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Formatting data...</span></span>
<span><span class="co">#&gt; Estimating size factors... (Max ETA: ~160.7s)</span></span>
<span><span class="co">#&gt; Finised in 144s!</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">size_factors_2</span><span class="op">)</span></span>
<span><span class="co">#&gt;        Ctrl1        Ctrl2        Ctrl3         LPS1         LPS2         LPS3 </span></span>
<span><span class="co">#&gt;  0.095303742  0.005168706 -0.083965484  0.073874335 -0.008215053  0.087963097 </span></span>
<span><span class="co">#&gt;          NE1          NE2          NE3       NELPS1       NELPS2       NELPS3 </span></span>
<span><span class="co">#&gt; -0.044352619  0.055490226  0.123255579 -0.181307332 -0.149171063 -0.025856786</span></span></code></pre></div>
<p>Indeed the estimates remain largely the same.</p>
</div>
<div class="section level3">
<h3 id="running-deseq">Running DESeq<a class="anchor" aria-label="anchor" href="#running-deseq"></a>
</h3>
<p>Once the size factors are inserted into the <code>DESeqDataSet</code>
object, the analysis then proceeds normally.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Constructing DESeqDataSet object</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html" class="external-link">DESeqDataSetFromMatrix</a></span><span class="op">(</span></span>
<span>    countData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>    colData <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    design <span class="op">=</span> <span class="va">design_formula</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">sizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="va">size_factors_2</span><span class="op">)</span> <span class="co"># Insert size factors</span></span>
<span></span>
<span><span class="co"># Run analysis</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html" class="external-link">DESeq</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span><span class="co">#&gt; using pre-existing size factors</span></span>
<span><span class="co">#&gt; estimating dispersions</span></span>
<span><span class="co">#&gt; gene-wise dispersion estimates</span></span>
<span><span class="co">#&gt; mean-dispersion relationship</span></span>
<span><span class="co">#&gt; final dispersion estimates</span></span>
<span><span class="co">#&gt; fitting model and testing</span></span>
<span></span>
<span><span class="co"># Print results</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu">DESeq2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html" class="external-link">results</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; log2 fold change (MLE): condition nelps vs control </span></span>
<span><span class="co">#&gt; Wald test p-value: condition nelps vs control </span></span>
<span><span class="co">#&gt; DataFrame with 57132 rows and 6 columns</span></span>
<span><span class="co">#&gt;                       baseMean log2FoldChange     lfcSE       stat      pvalue</span></span>
<span><span class="co">#&gt;                      &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ENSMUSG00000028180 1.94379e+03     -0.8031370 0.0948941 -8.4635057 2.59458e-17</span></span>
<span><span class="co">#&gt; ENSMUSG00000028182 2.82072e+01     -0.9036829 0.5053360 -1.7882814 7.37306e-02</span></span>
<span><span class="co">#&gt; ENSMUSG00000028185 3.36083e-01     -0.0567436 4.4073563 -0.0128748 9.89728e-01</span></span>
<span><span class="co">#&gt; ENSMUSG00000028184 1.03770e+04      0.7491799 0.0647543 11.5695750 5.87731e-31</span></span>
<span><span class="co">#&gt; ENSMUSG00000028187 9.79971e+02     -0.1955160 0.1124985 -1.7379436 8.22207e-02</span></span>
<span><span class="co">#&gt; ...                        ...            ...       ...        ...         ...</span></span>
<span><span class="co">#&gt; ENSMUSG00000106108    10.66611      6.7370233 1.4341914   4.697437 2.63447e-06</span></span>
<span><span class="co">#&gt; ENSMUSG00000042675  1583.23329     -0.0525418 0.0782159  -0.671753 5.01741e-01</span></span>
<span><span class="co">#&gt; ENSMUSG00000036959  1428.91454      0.5311799 0.0834599   6.364493 1.95936e-10</span></span>
<span><span class="co">#&gt; ENSMUSG00000036958     0.66323     -3.9148515 4.2776350  -0.915191 3.60092e-01</span></span>
<span><span class="co">#&gt; ENSMUSG00000042678     5.59561     -5.6136803 1.7045407  -3.293368 9.89947e-04</span></span>
<span><span class="co">#&gt;                           padj</span></span>
<span><span class="co">#&gt;                      &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ENSMUSG00000028180 2.37920e-16</span></span>
<span><span class="co">#&gt; ENSMUSG00000028182 1.43474e-01</span></span>
<span><span class="co">#&gt; ENSMUSG00000028185          NA</span></span>
<span><span class="co">#&gt; ENSMUSG00000028184 9.03694e-30</span></span>
<span><span class="co">#&gt; ENSMUSG00000028187 1.57786e-01</span></span>
<span><span class="co">#&gt; ...                        ...</span></span>
<span><span class="co">#&gt; ENSMUSG00000106108 1.05146e-05</span></span>
<span><span class="co">#&gt; ENSMUSG00000042675 6.53594e-01</span></span>
<span><span class="co">#&gt; ENSMUSG00000036959 1.14561e-09</span></span>
<span><span class="co">#&gt; ENSMUSG00000036958          NA</span></span>
<span><span class="co">#&gt; ENSMUSG00000042678 2.84207e-03</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Todd McCready.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
