<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Implementation of disize • disize</title>
<!-- katex math --><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous">
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script><script src="../katex-auto.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Implementation of disize">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">disize</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.30</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/disize.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/benchmarking.html">Benchmarking against DESeq2 and edgeR</a></li>
    <li><a class="dropdown-item" href="../articles/comparing-defs.html">Comparing batch definitions</a></li>
    <li><a class="dropdown-item" href="../articles/implementation.html">Implementation of disize</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/toddmccready/disize/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Implementation of disize</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/toddmccready/disize/blob/main/vignettes/implementation.Rmd" class="external-link"><code>vignettes/implementation.Rmd</code></a></small>
      <div class="d-none name"><code>implementation.Rmd</code></div>
    </div>

    
    
<p>Internally, <code>disize</code> uses Stan to fit a Bayesian model
that jointly estimates the effect of covariates (structured according to
<code>design_formula</code>) on gene expression <em>and</em> any
confounding batch-effects:</p>
<p><span class="math display">\begin{aligned}
    \mathbf{y}_g &amp;\sim \text{NegBinom}(\mathbf{\mu}_g, \phi) \\
    \log \mathbf{\mu}_g &amp;= \mathbf{\alpha}_g + \mathbf{X}
\mathbf{\beta}_g + \mathbf{Z}\mathbf{b}_g + \mathbf{o} \\
        &amp;= \mathbf{\alpha}_g + \mathbf{X} \mathbf{\beta}_g +
\mathbf{Z}\mathbf{b}_g + (\mathbf{B} \mathbf{s}) \\
    \mathbf{b}_g &amp;\sim \text{Normal}(\mathbf{0}, \mathbf{G}_g)
\end{aligned}</span></p>
<p>Where <span class="math inline">\mathbf{y}_g</span> denotes the
vector of counts of a gene <span class="math inline">g</span> for all
observations, which is realized from the negative binomial parameterized
by the effect of the covariates (<span class="math inline">\mathbf{\alpha}_g + \mathbf{X} \mathbf{\beta}_g +
\mathbf{Z}\mathbf{b}_g</span>) and any batch-effects (<span class="math inline">\mathbf{B} \mathbf{s}</span>).</p>
<p>The experimental design is specified by an R formula
(<code>design_formula</code>) that constructs a “fixed-effects” model
matrix <span class="math inline">\mathbf{X}</span> (without an
intercept) and a “random-effects” model matrix <span class="math inline">\mathbf{Z}</span>; this by itself is a regular
GLMM.</p>
<p>The confounding batch-effect is essentially an unknown offset <span class="math inline">\mathbf{o} = \mathbf{B} \mathbf{s}</span>, where
<span class="math inline">\mathbf{B}</span> specifies the batch
membership for each observation and <span class="math inline">\mathbf{s}</span> contains the “size factors” which
scale the true magnitude of expression.</p>
<p>At its face, however, this model should not be identifiable for most
experimental designs (often the batch ID results in too many free
parameters such as when it is identical to a measured covariate). This
identifiability issue is overcome by constraining <span class="math inline">\mathbf{s}</span> and assuming only a fraction of
features are significantly affected by the covariates measured in the
experiment; in other words, the estimated coefficients <span class="math inline">\mathbf{\beta}_g, \mathbf{b}_g</span> (excluding the
intercept) are <em>sparse</em> across genes.</p>
<p>This assumption is encoded in the model by placing distinct horseshoe
priors on each of the model coefficients (both the fixed- and
random-effects). This allows the priors to be learned independently of
each other using the large number of features measured in RNAseq
experiments.</p>
<div class="section level2">
<h2 id="estimation">Estimation<a class="anchor" aria-label="anchor" href="#estimation"></a>
</h2>
<p>Since the posterior distribution for <span class="math inline">\mathbf{s}</span> is heavily concentrated for
typical datasets with thousands of features, it is sufficient to quickly
provide a point estimate of <span class="math inline">\mathbf{s}</span>
and delegate estimation of the model coefficients to existing tools like
<code>DESeq2</code> or <code>edgeR</code> (only replacing the small
normalization step of their workflows).</p>
<p><code>disize</code> uses Stan’s <a href="https://mc-stan.org/docs/reference-manual/optimization.html" class="external-link">L-BFGS
optimization algorithm</a> to find the model’s <em>maximum a
posteriori</em> (MAP) for <span class="math inline">\mathbf{s}</span>.
We end up doing this in fewer iterations than needed for all parameters
to converge by using a heuristic to guess how long the procedure should
run for; this is followed up with a diagnostic to ensure the size
factors have converged.</p>
</div>
  </main>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Todd McCready.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
